### ISSUE TITLE
Trin should use only one database

### ISSUE THREAD

**Issue by mrferris** [ethereum/trin#240]
Currently, Trin uses two databases: RocksDB and SQLite. Originally only RocksDB was used. It was chosen because it's a very performant key-value store, and the portal network is a big key-value store. So that made sense. SQLite was later added once it became clear that RocksDB currently does not appear to be able to sort its set of data by an arbitrary function, which is necessary for finding the farthest piece of data stored.

There are multiple reasons that Trin should use one database instead of using both RocksDB and SQLite as it does currently.

1. Performance, as pointed out by @carver here: https://github.com/ethereum/trin/issues/91
2. Complexity involved with keeping DBs in-sync with each other in case of failure, as pointed out by @jacobkaufmann here: https://github.com/ethereum/trin/pull/230#discussion_r798106431

Recent discussion about this with @njgheorghita led to the idea of using only SQLite, and then re-evaluating our DB strategy in the future if SQLite is measured to be any kind of significant bottle-neck upon deployment of the network.

**Comment by njgheorghita** [comment]
I agree that using two databases might not be ideal. But a couple of other thoughts come to mind now. 
- What are the consequences of out-of-sync dbs?
- How often can we reasonably estimate that dbs will fall out of sync?
- Could we just provide some util function that will clear the metadb, iterate over rocksdb, & reseed the metadb should they fall out of sync?
Ultimately, I wouldn't consider this that high of a priority. Performance does not appear to me to be one of our primary concerns. But as we get more features shipped, I'd love to see some benchmarks comparing a few of the possible strategies

**Comment by mrferris** [comment]
On the one hand I see your point, on the other hand I think that leaving two DBs is actually more work for right now because we should then consider the excellent questions that you brought up above. As opposed to implementing the key/value get/set/delete functionality in SQLite that RocksDB is currently doing which is very straightforward.

RocksDB also makes up 54% of the crate binary data in the Trin binary according to building master currently: `Downloaded 115 crates (15.0 MB) in 3.39s (largest was `librocksdb-sys` at 8.1 MB)`, which also might be a reason to remove it until we find that we need it.

**Comment by njgheorghita** [comment]
Ok, well yeah I guess I'm sold. It doesn't seem that complex of a change, and can easily (as far as I can tell) be improved / refactored for performance when performance becomes high-priority. But, I guess there's one last blocker. `TrieDB`, the library @carver built for the state network to use relies on `rocksdb` [here](https://github.com/ethereum/trin/blob/master/trin-state/src/network.rs#L31). On one hand, we're currently targeting the history network, so maybe we drop that import until we're focused on the state network, to benefit from dropping the rocksdb dependency. On the other, it does seem inevitable that we'll eventually end up using `rocksdb` in some manner, is it smart to move over to sqlite entirely now, even for some temporary short-term gains. 

 `"it does seem inevitable that we'll eventually end up using `rocksdb` in some manner,"` - Ok, I guess this is also an assumption, but one I'm somewhat comfortable with.

**Comment by mrferris** [comment]
Good point on `TrieDB`, that definitely changes things. Let's take a closer look at whether the necessary functionality mentioned above is possible with RocksDB.

**Comment by KolbyML** [comment]
https://github.com/btrask/sqlheavy looking at projects like this it must be possible

(this is drop in replacement for sqlite using for example rocksdb as storage engine)

**Comment by KolbyML** [comment]
> SQLite was later added once it became clear that RocksDB currently does not appear to be able to sort its set of data by an arbitrary function, which is necessary for finding the farthest piece of data stored.

From looking into it a bit maybe I am missing something aggregates like ``TOTAL, COUNT`` iterate over the whole DB in SQLite hence they are O(n) ``TOTAL_DATA_SIZE_QUERY``/``TOTAL_ENTRY_COUNT_QUERY`` Could be replaced iterating over our whole rocksDB (which would be exactly what SQLite is doing with its storage engine, except with the overhead of SQL).

``CONTENT_KEY_LOOKUP_QUERY``/``CONTENT_SIZE_LOOKUP_QUERY`` we can use Rocks get?

``PAGINATE_QUERY`` https://stackoverflow.com/questions/63318315/sorting-by-value-in-rocksdb can be done with this and a rocksdb ``snapshot`?

``XOR_FIND_FARTHEST_QUERY`` this is just a ORDER BY DESC which would just be ``n*log(n)`` which RocksDB iterator then a sort? This could easily be implemented in n*logn

If we only use rocks DB we won't need ``CREATE_QUERY``, ``INSERT_QUERY``, ``DELETE_QUERY`` since we already have that

### TL:DR
Looking into it, it seems very possible to implement this functionality with RocksDB which is simply just a faster storage engine then SQLite, Our implementations for the ``replacements`` for the functions above would be the same time complexity (it isn't like sqlite has some special unknown sauce for Data Structures and Algo's). 

So the main benefits of this would be 
1. decreasing PortalStorage complexity since we would only need 1 DB again
2. none of the potential issues of using 2 key-store databases at once
3. lowering disk usage (we wouldn't have do everything twice)
4. responses to get what we need from PortalStorage since RocksDB is a faster engine

**Comment by KolbyML** [comment]
Ah ok so purge/paginate use sqlite for content_key, rocksdb doesn't include that right now. content_key can derive content_id used for xor.

So if we used content_key for rocksdb key then there would be no issue. For get() in rocksdb we already derive content_id there. Another solution would be union(content_key, data).

But maybe I should try and bench mark if total is faster doing it like this. (maybe a index is magic)

😵‍💫I need to look into things more before I talk.
